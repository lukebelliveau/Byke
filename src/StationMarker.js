import React from 'react';
import MapView from 'react-native-maps';

import Circle from './Circle';

const Marker = ({ coordinate, stationName, availableBikes }) =>
  <MapView.Marker
    coordinate={coordinate}
    title={stationName}
    description={`${availableBikes} bikes available`}
  >
    <Circle color={availableBikes > 5 ? 'green' : 'red'} />
  </MapView.Marker>;

export default Marker;
