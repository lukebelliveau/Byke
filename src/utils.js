// @flow
import { Location, Station } from './Types';

const computeRegionThatFitsAllPoints = (landmarks: Location) => {
  let top = -100;
  let bottom = 100;
  let left = 100;
  let right = -100;

  landmarks.forEach(landmark => {
    if (landmark.latitude > right) right = landmark.latitude;
    if (landmark.latitude < left) left = landmark.latitude;
    if (landmark.longitude > top) top = landmark.longitude;
    if (landmark.longitude < bottom) bottom = landmark.longitude;
  });

  const longitude = (top - bottom) / 2 + bottom;
  const latitude = (right - left) / 2 + left;
  const latitudeDelta = right - left + 0.005;
  const longitudeDelta = top - bottom + 0.005;

  return {
    latitude,
    longitude,
    latitudeDelta,
    longitudeDelta,
  };
};

const findClosestStation = (
  currentLocation: Location,
  stations: Array<Station>
) => {
  let closestFound = {
    station: null,
    distance: Number.POSITIVE_INFINITY,
  };

  stations.forEach(station => {
    const lateralDistance = Math.abs(
      currentLocation.latitude - station.latitude
    );
    const longitudalDistance = Math.abs(
      currentLocation.longitude - station.longitude
    );
    const distance = lateralDistance + longitudalDistance / 2;

    if (distance < closestFound.distance)
      closestFound = { station: station, distance: distance };
  });

  return closestFound.station;
};

export default {
  computeRegionThatFitsAllPoints,
  findClosestStation,
};
