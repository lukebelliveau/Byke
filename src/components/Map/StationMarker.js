// @flow
import React from 'react';
import MapView from 'react-native-maps';
import { Keyboard } from 'react-native';
import utils from '../../utils';

import Circle from '../shared/Circle';

const Marker = ({
  coordinate,
  stationName,
  currentLocation,
  availableBikes,
  availableDocks,
}: {
  coordinate: Object,
  stationName: string,
  availableBikes: number,
}) => {
  return (
    <MapView.Marker
      coordinate={coordinate}
      title={stationName}
      description={`${availableBikes} bikes available, ${availableDocks} docks available`}
      onPress={Keyboard.dismiss}
      onCalloutPress={() =>
        utils.displayNavigationAlert(
          currentLocation,
          stationName,
          coordinate,
          availableBikes,
          availableDocks
        )}
    >
      <Circle color={availableBikes > 5 ? 'green' : 'red'} />
    </MapView.Marker>
  );
};

export default Marker;
