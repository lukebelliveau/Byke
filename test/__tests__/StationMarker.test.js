import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Keyboard } from 'react-native';

import StationMarker from '../../src/components/Map/StationMarker';
import utils from '../../src/utils';

const coordinate = { latitude: 50, longitude: 50 };

it('opens Google Maps with origin and destination coordinates', () => {
  utils.openDirections = jest.fn();

  const currentLocation = { latitude: 50, longitude: 50 };
  const destination = { latitude: 100, longitude: 100 };
  const marker = shallow(
    <StationMarker
      currentLocation={currentLocation}
      coordinate={destination}
    />
  );

  marker.simulate('calloutPress');

  expect(utils.openDirections).toBeCalledWith(currentLocation, destination);
});

it('should be green when more than 5 bikes are available', () => {
  const marker = renderer
    .create(
      <StationMarker
        coordinate={coordinate}
        stationName="station"
        availableBikes={8}
        availableDocks={10}
      />
    )
    .toJSON();

  expect(marker).toMatchSnapshot();
});

it('should be red when 5 or less bikes are available', () => {
  const marker = renderer
    .create(
      <StationMarker
        coordinate={coordinate}
        stationName="station"
        availableBikes={4}
        availableDocks={10}
      />
    )
    .toJSON();

  expect(marker).toMatchSnapshot();
});

it('dismisses the keyboard when pressed', () => {
  Keyboard.dismiss = jest.fn();
  const marker = shallow(
    <StationMarker
      coordinate={coordinate}
      stationName="Lincoln Park"
      availableBikes={6}
    />
  );

  marker.simulate('press');

  expect(Keyboard.dismiss.mock.calls.length).toBeGreaterThan(0);
});

/*
 * CONFIG (react-test-renderer & react-native-maps don't play nice)
 * see https://github.com/airbnb/react-native-maps/issues/889
 * */
jest.mock('react-native-maps', () => {
  const React = require.requireActual('react');
  const MapView = require.requireActual('react-native-maps');

  class MockCallout extends React.Component {
    render() {
      return React.createElement('Callout', this.props, this.props.children);
    }
  }

  class MockMarker extends React.Component {
    render() {
      return React.createElement('Marker', this.props, this.props.children);
    }
  }

  class MockMapView extends React.Component {
    render() {
      return React.createElement('MapView', this.props, this.props.children);
    }
  }

  MockCallout.propTypes = MapView.Callout.propTypes;
  MockMarker.propTypes = MapView.Marker.propTypes;
  MockMapView.propTypes = MapView.propTypes;
  MockMapView.Marker = MockMarker;
  MockMapView.Callout = MockCallout;
  return MockMapView;
});
