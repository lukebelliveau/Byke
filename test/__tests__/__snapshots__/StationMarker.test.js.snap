// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be green 1`] = `
<Marker
  coordinate={
    Object {
      "latitude": 50,
      "longitude": 50,
    }
  }
  description="8 bikes available, 10 docks available"
  onCalloutPress={[Function]}
  onPress={[Function]}
  title="station"
>
  <View
    style={
      Object {
        "backgroundColor": "green",
        "borderRadius": 10,
        "height": 15,
        "width": 15,
      }
    }
  />
</Marker>
`;

exports[`should be green when more than 5 bikes are available 1`] = `
<Marker
  coordinate={
    Object {
      "latitude": 50,
      "longitude": 50,
    }
  }
  description="8 bikes available, 10 docks available"
  onCalloutPress={[Function]}
  onPress={[Function]}
  title="station"
>
  <View
    style={
      Object {
        "backgroundColor": "green",
        "borderRadius": 10,
        "height": 15,
        "width": 15,
      }
    }
  />
</Marker>
`;

exports[`should be red 1`] = `
<Marker
  coordinate={
    Object {
      "latitude": 50,
      "longitude": 50,
    }
  }
  description="4 bikes available, 10 docks available"
  onCalloutPress={[Function]}
  onPress={[Function]}
  title="station"
>
  <View
    style={
      Object {
        "backgroundColor": "red",
        "borderRadius": 10,
        "height": 15,
        "width": 15,
      }
    }
  />
</Marker>
`;

exports[`should be red when 5 or less bikes are available 1`] = `
<Marker
  coordinate={
    Object {
      "latitude": 50,
      "longitude": 50,
    }
  }
  description="4 bikes available, 10 docks available"
  onCalloutPress={[Function]}
  onPress={[Function]}
  title="station"
>
  <View
    style={
      Object {
        "backgroundColor": "red",
        "borderRadius": 10,
        "height": 15,
        "width": 15,
      }
    }
  />
</Marker>
`;
